### PR TITLE
LIDO-3077 : 2D Array

### PR DESCRIPTION
**Pull request by marhong2**
Updated based upon comments and added the ability to call functions with in arrays.

### CODE REVIEW

File: grammar/parsing.js [hunk 1]
```diff
@@ -206,12 +206,80 @@ class Parsing extends EmbeddedActionsParser {
...
         });
 
+        $.RULE('multiArray', () => {
+            const multiArr = [[]];
+            let singleArr = null;
+
+            const updateRow = (multiArr, newArr) => {
+                if (multiArr[0] === null)
+                    multiArr.push([])
+                for(let newArrIndex = 0; newArrIndex < newArr.length; newArrIndex++) {
```
Comment (tyleradams) [inline grammar/parsing.js:216]: Use map over for loops
newArr.map(a => {
})

**Review by tyleradams — APPROVED** [review]
Once comments are addressed, feel free to merge

File: grammar/parsing.js [hunk 2]
```diff
@@ -206,12 +206,80 @@ class Parsing extends EmbeddedActionsParser {
...
+                    let currVal = newArr[newArrIndex];
+                    if (!Array.isArray(currVal)) {
+                        multiArr[0].push(currVal);
+                        continue;
+                    }
+                    if (multiArr.length <= newArrIndex)
+                            multiArr.push([]);
```
Comment (jmorganmartin) [inline grammar/parsing.js:223]: I think there's a bug here. This `multiArr.push([])` does not guarantee that `multiArr[newArrIndex]` will be defined on L225.  To do that, you'd need to do a loop:

```js
  while (multiArr.length <= newArrIndex) { 
    multiArr.push([]);
  }
```
***But*** I don't think that's what we ultimately want here either.

Take a case where the input `newArr` contains a lot of non-array values:

```js
  newArr = [0, 1, 2, 3, 4, [5, 6, 7, 8]];
```

What is the intended value of `multiArr` after calling this `updateRow` function? In my head, I'm calculating that `multiArr` would look like this when the code hits L222:

```js
  [[0, 1, 2, 3, 4]];
```

Then, after L223:

```js
  [[0, 1, 2, 3, 4], []];
```

The problem now is that `newArrIndex` will be set to `5`, but `multiArr` only has a length of `2`. So on L225, we try to `multiArr[5].push()`, but `multiArr[5]` is not yet defined, so an error is thrown. If we change that `if` to a `while`, we'll avoid the error, but end up with empty values in `multiArr`, e.g.:

`multiArr` after while loop:
```js
[[0, 1, 2, 3, 4], [], [], [], [], []];
```

Final `multiArr`:

```js
[[0, 1, 2, 3, 4], [], [], [], [], [5, 6, 7, 8]];
```

TLDR; I can't tell the exact intention of this function, but the current implementation appears to be prone to bugs/errors. If possible, see if this function can be extracted and write some tests to ensure it's working as intended.

**Review by jmorganmartin — DISMISSED** [review]
It looks like the additional tests necessitated a new `updateRow` function (I don't recall seeing this in previous commit), and I have a few different concerns with that function.

Also, please try not to squash intermediate commits that aren't trivial/errant. Since there is only one commit on this branch, we can't see the changes since the last review.

File: grammar/parsing.js [hunk 3]
```diff
@@ -206,12 +206,80 @@ class Parsing extends EmbeddedActionsParser {
             return result;
         });
 
+        $.RULE('multiArray', () => {
+            const multiArr = [[]];
+            let singleArr = null;
+
+            const updateRow = (multiArr, newArr) => {
+                if (multiArr[0] === null)
```
Comment (jmorganmartin) [inline grammar/parsing.js:214]: Not sure if this check is necessary given the initial value assigned to `multiArr` on L210. However, if this function is to be reused in other situations, it's good to have some sort of sanity check.

The bigger problem is I'm not sure it's going to work as intended.  L214 and L215 will do exactly one thing:
if `multiArr = [null]` (could have more than one element; the first one just has to be exactly `null`), it will update it to be `[null, []]` (again, could have more than two elements, but the last one is gonna be an empty array).

Note that it the condition `multiArr[0] === null` will _not_ be met by `multiArr = []` because of the `===`. If you just want to check if the first value is empty, you could potentially do `multiArr[0] == null`, as `== null` matches exactly two things `null` and `undefined`.

But, ultimately, why not what is used below?: 

```js
if (!Array.isArray(multiArr[0]))
```

Note: if we are concerned that `multiArr` might not be an array at all (in which case, checking `multiArr[0]` will throw an undefined error), we'd actually need an additional check for that: 

```
if (!Array.isArray(multiArr)) { 
  // bad input. throw error since we assume multiArr is a const, so it can't be redefined as an array 
}
```

File: grammar/parsing.js [hunk 4]
```diff
@@ -206,12 +206,80 @@ class Parsing extends EmbeddedActionsParser {
...
+                    multiArr.push([])
+                for(let newArrIndex = 0; newArrIndex < newArr.length; newArrIndex++) {
+                    let currVal = newArr[newArrIndex];
+                    if (!Array.isArray(currVal)) {
+                        multiArr[0].push(currVal);
+                        continue;
```
Comment (jmorganmartin) [inline grammar/parsing.js:220]: Note: inside an `Array.map()`, you cannot use `continue`; but you can use `return` to short-circuit out of a given loop.

File: grammar/dependency/utils.js [hunk 5]
```diff
@@ -286,6 +286,41 @@ class Utils {
...
+    * becomes:
+    * multiRow = [[1,2], [3,4]]
+    * on the spread sheet:
+    * 1 2
+    * 3 4
+    */
+    updateRow = (multiArr, inputArr) => {
+        if(!Array.isArray(multiArr))
+            throw "ERROR: multiArray is not array";
+        if (multiArr[0] === undefined)
```
Comment (tyleradams) [inline grammar/dependency/utils.js:307]: check length of multiArr

File: grammar/parsing.js [hunk 6]
```diff
@@ -228,13 +279,10 @@ class Parsing extends EmbeddedActionsParser {
...
                 }
             });
-
-            $.CONSUME(CloseCurlyParen);
-
             return $.ACTION(() => this.utils.toArray(arr));
         });
-
-        /**
+
```
Comment (tyleradams) [inline grammar/parsing.js:284]: no trailing whitespace

File: grammar/utils.js [hunk 7]
```diff
@@ -403,6 +403,41 @@ class Utils {
...
+    * such that the columns of multiArr and inputArr align
+    *  
+    * updateRow([[1], [3]], [[2], [3]])
+    * becomes:
+    * multiRow = [[1,2], [3,4]]
+    * on the spread sheet:
+    * 1 2
+    * 3 4
+    */
+    updateRow = (multiArr, inputArr) => {
```
Comment (tyleradams) [inline grammar/utils.js:421]: Why is this code duplicated?

Comment (marhong2) [reply grammar/utils.js:421]: The folders /grammar and /test/grammar are identical and both contain a utils file (where updateRow resides) and a hooks file (imports utils and parses). both copies of Hooks and utils are identical. I assumed one of the files was for testing purposes and the other for deployment. Thus I copied the function into both.

Comment (marhong2) [reply grammar/utils.js:421]: TLDR, I removed updateRow from both independently and hit errors

Comment (tyleradams) [reply grammar/utils.js:421]: Put code in FormulaParser and import from there

Comment (marhong2) [reply grammar/utils.js]: I did but depParser threw errors b/c it also calls the same parser. Thus, I added updateRow to it too. I tried to find a single file to put depParser into, but since the only values passed into Parser is DepParser/FormulaParser and their respective utils, I do not think it is possible to have updateRow in one place, unless you want to combine util files.

Comment (tyleradams) [reply grammar/utils.js]: I don't understand why the same parser impacts importing from a file, but let's just merge this in and follow up right after.
